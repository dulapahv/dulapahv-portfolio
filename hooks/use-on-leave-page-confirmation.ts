'use client';

import { useEffect, useRef } from 'react';
import { usePathname, useRouter } from 'next/navigation';

const MESSAGE = 'Changes you made may not be saved.';

export function useOnLeavePageConfirmation(shouldConfirm: boolean) {
  const pathname = usePathname();
  const router = useRouter();

  const isInitialMount = useRef(true);

  // Handle refresh or close tab/window
  useEffect(() => {
    if (!shouldConfirm) return;

    function beforeUnload(e: BeforeUnloadEvent) {
      e.preventDefault();
      e.returnValue = '';
    }

    window.addEventListener('beforeunload', beforeUnload);
    return () => window.removeEventListener('beforeunload', beforeUnload);
  }, [shouldConfirm]);

  // Handle back/forward browser buttons
  useEffect(() => {
    if (!shouldConfirm) return;

    const handleBackButton = () => {
      if (!window.confirm(MESSAGE)) history.pushState(null, '', window.location.href);
      else {
        window.removeEventListener('popstate', handleBackButton);
        history.back();
        // When user clicks back after refreshing the page with unsaved changes
        setTimeout(() => {
          if (window.location.pathname === pathname) router.back();
        }, 0);
      }
    };

    // Push initial state to history to detect back navigation
    if (isInitialMount.current) {
      history.pushState(null, '', window.location.href);
      isInitialMount.current = false;
    }

    window.addEventListener('popstate', handleBackButton);
    return () => window.removeEventListener('popstate', handleBackButton);
  }, [shouldConfirm, pathname]);

  // Handle Next.js Link navigation
  useEffect(() => {
    if (!shouldConfirm) return;

    const handleClick = (e: MouseEvent) => {
      const target = e.target as HTMLElement;
      const anchor = target.closest('a');

      if (anchor && anchor.href && anchor.href.startsWith(window.location.origin)) {
        const targetUrl = new URL(anchor.href);
        const currentPath = pathname;

        if (targetUrl.pathname !== currentPath) {
          if (!window.confirm(MESSAGE)) {
            e.preventDefault();
            e.stopPropagation();
          }
        }
      }
    };

    document.addEventListener('click', handleClick, true);
    return () => document.removeEventListener('click', handleClick, true);
  }, [shouldConfirm, pathname]);
}
