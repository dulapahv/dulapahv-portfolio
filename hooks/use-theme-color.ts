"use client";

import { useEffect } from 'react';

import { useTheme } from 'next-themes';

export function useThemeColor() {
  const { resolvedTheme } = useTheme();

  useEffect(() => {
    // Get the theme color meta tag
    const themeColorMeta = document.querySelector('meta[name="theme-color"]');

    if (themeColorMeta) {
      // Update the color based on theme
      themeColorMeta.setAttribute(
        'content',
        resolvedTheme === 'dark' ? '#0a0a0a' : '#eef1f7',
      );
    }
  }, [resolvedTheme]);
}
