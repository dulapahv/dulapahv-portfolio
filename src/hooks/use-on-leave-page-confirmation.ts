"use client";

import { useEffect } from "react";

export function useOnLeavePageConfirmation(isDirty: boolean) {
  useEffect(() => {
    function beforeUnload(e: BeforeUnloadEvent) {
      if (!isDirty) return;
      e.preventDefault();
    }

    window.addEventListener("beforeunload", beforeUnload);

    return () => {
      window.removeEventListener("beforeunload", beforeUnload);
    };
  }, [isDirty]);
}
