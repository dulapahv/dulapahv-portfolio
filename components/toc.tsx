'use client';

/* Honestly, this is a bit of a mess. */
import { useCallback, useEffect, useRef, useState } from 'react';
import Link from 'next/link';

import {
  ArrowCircleUpIcon,
  CaretDownIcon,
} from '@phosphor-icons/react/dist/ssr';
import { AnimatePresence, motion } from 'motion/react';

import { cn } from '@/lib/utils';
import { useMediaQuery } from '@/hooks/use-media-query';

interface TOCItem {
  id: string;
  text: string;
  level: number;
}

export const TableOfContents = () => {
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [tocItems, setTocItems] = useState<TOCItem[]>([]);
  const [activeId, setActiveId] = useState<string>('');
  const [lockActiveId, setLockActiveId] = useState<boolean>(false);
  const [isCollapsed, setIsCollapsed] = useState<boolean>(true);
  const [showScrollTop, setShowScrollTop] = useState<boolean>(false);
  const isDesktop = useMediaQuery('(min-width: 1280px)'); // xl breakpoint
  const tocRef = useRef<HTMLDivElement>(null);
  const linksRef = useRef<(HTMLAnchorElement | null)[]>([]);

  // Handle scroll visibility for scroll to top button
  useEffect(() => {
    const handleScroll = () => {
      // Show button after scrolling down 300px
      setShowScrollTop(window.scrollY > 300);
    };

    window.addEventListener('scroll', handleScroll);
    return () => window.removeEventListener('scroll', handleScroll);
  }, []);

  useEffect(() => {
    const updateTOC = () => {
      const headings = Array.from(document.querySelectorAll('h2[id], h3[id]'));
      const newTocItems = headings.map((el) => ({
        id: el.id,
        text: el.textContent || '',
        level: parseInt(el.tagName[1]),
      }));
      setTocItems(newTocItems);
      setIsLoading(false);

      // Reset links ref array
      linksRef.current = new Array(newTocItems.length).fill(null);

      // Only set initial active ID if we don't have one yet
      setActiveId((currentActiveId) => {
        if (currentActiveId === '' && newTocItems.length > 0) {
          const hash = window.location.hash.replace('#', '');
          if (hash && newTocItems.some((item) => item.id === hash)) {
            return hash;
          } else {
            return newTocItems[0].id;
          }
        }
        return currentActiveId;
      });
    };

    updateTOC();

    const mdxContainer = document.querySelector('article');
    if (!mdxContainer) return;

    const observer = new MutationObserver((mutations) => {
      // Check if the mutation is just a button state change
      const isButtonChange = mutations.some((mutation) => {
        const target = mutation.target as HTMLElement;
        return (
          target.closest('button') ||
          (mutation.type === 'attributes' &&
            mutation.attributeName === 'disabled') ||
          (mutation.type === 'childList' && target.querySelector('svg'))
        );
      });

      // Only update TOC for structural changes, not button state changes
      if (!isButtonChange) {
        updateTOC();
      }
    });

    observer.observe(mdxContainer, {
      childList: true,
      subtree: true,
      attributes: true,
      attributeFilter: ['disabled'],
    });

    const sectionObserver = new IntersectionObserver(
      (entries) => {
        entries.forEach((entry) => {
          const id = entry.target.getAttribute('id');
          if (entry.isIntersecting && !lockActiveId) {
            setActiveId(id ?? '');
          }
        });
      },
      { rootMargin: '0px 0px -80% 0px', threshold: 1 },
    );

    const headings = Array.from(document.querySelectorAll('h2[id], h3[id]'));
    headings.forEach((heading) => {
      // Make headings focusable for proper tab navigation
      if (!heading.hasAttribute('tabindex')) {
        heading.setAttribute('tabindex', '-1');
      }
      sectionObserver.observe(heading);
    });

    return () => {
      observer.disconnect();
      sectionObserver.disconnect();
    };
  }, [lockActiveId]);

  const scrollToElement = useCallback(
    (id: string) => {
      const element = document.getElementById(id);
      if (element) {
        element.scrollIntoView({ block: 'start' });
        window.history.replaceState(null, '', `#${id}`);
        setActiveId(id);
        setLockActiveId(true);
        setTimeout(() => setLockActiveId(false), 500);

        // Focus the target heading for screen readers after scroll completes
        element.focus({ preventScroll: true });
      }
    },
    [isDesktop],
  );

  const scrollToTop = useCallback(() => {
    window.scrollTo({ top: 0, behavior: 'smooth' });
    window.history.replaceState(null, '', window.location.pathname);
  }, []);

  const handleClick = (e: React.MouseEvent<HTMLAnchorElement>, id: string) => {
    e.preventDefault();
    scrollToElement(id);
  };

  const handleKeyDown = (
    e: React.KeyboardEvent<HTMLAnchorElement>,
    id: string,
    index: number,
  ) => {
    switch (e.key) {
      case 'Enter':
      case ' ':
        e.preventDefault();
        scrollToElement(id);
        break;
      case 'ArrowDown':
        e.preventDefault();
        e.stopPropagation();
        const nextIndex = Math.min(index + 1, tocItems.length - 1);
        linksRef.current[nextIndex]?.focus();
        break;
      case 'ArrowUp':
        e.preventDefault();
        e.stopPropagation();
        const prevIndex = Math.max(index - 1, 0);
        linksRef.current[prevIndex]?.focus();
        break;
      case 'Home':
        e.preventDefault();
        e.stopPropagation();
        linksRef.current[0]?.focus();
        break;
      case 'End':
        e.preventDefault();
        e.stopPropagation();
        const lastIndex = tocItems.length - 1;
        linksRef.current[lastIndex]?.focus();
        break;
    }
  };

  // Show loading state while scanning for headings
  if (isLoading) {
    return (
      <div
        className="border-border bg-background-elevated/50 text-foreground-muted flex w-full
          items-center rounded-md border px-4 py-3 text-sm backdrop-blur-sm xl:hidden"
      >
        Loading table of contents...
      </div>
    );
  }

  // Don't render if there are no headings
  if (tocItems.length === 0) {
    return null;
  }

  // Desktop view - fixed sidebar
  if (isDesktop) {
    return (
      <nav
        className="fixed top-18 left-[calc(100vw/2+404px)] z-40 hidden w-56 xl:block"
        aria-label="Table of contents"
        role="navigation"
        ref={tocRef}
      >
        <h2 className="text-foreground px-3 text-sm font-semibold">
          On this page
        </h2>
        <div
          className="max-h-[calc(100vh-16rem)] overflow-y-auto py-2"
          role="group"
          aria-label="Page sections - use arrow keys to navigate"
        >
          <ul className="space-y-1" role="list">
            {tocItems.map((item, index) => (
              <li key={item.id} className="mb-0">
                <Link
                  ref={(el) => {
                    linksRef.current[index] = el;
                  }}
                  href={`#${item.id}`}
                  onClick={(e) => handleClick(e, item.id)}
                  onKeyDown={(e) => handleKeyDown(e, item.id, index)}
                  className={cn(
                    'group relative block rounded-md px-3 py-1 text-sm font-medium transition-all',
                    item.level === 3 ? 'pl-7' : '',
                    activeId === item.id
                      ? 'text-mirai-red'
                      : 'text-foreground-muted',
                    activeId !== item.id && 'hover:text-foreground',
                  )}
                  aria-current={activeId === item.id ? 'location' : undefined}
                  aria-describedby={
                    activeId === item.id
                      ? `current-section-${index}`
                      : undefined
                  }
                >
                  <span
                    className={cn(
                      'absolute top-1/2 left-0 h-full w-0.5 -translate-y-1/2 transition-all',
                      activeId === item.id
                        ? 'bg-mirai-red'
                        : 'bg-white/80 dark:bg-neutral-300/80',
                    )}
                    aria-hidden="true"
                  />
                  <span className="relative line-clamp-2">{item.text}</span>
                  {activeId === item.id && (
                    <span id={`current-section-${index}`} className="sr-only">
                      (current section)
                    </span>
                  )}
                </Link>
              </li>
            ))}
          </ul>
        </div>

        {/* Scroll to top button */}
        <AnimatePresence>
          {showScrollTop && (
            <motion.div
              initial={{ opacity: 0, y: 10 }}
              animate={{ opacity: 1, y: 0 }}
              exit={{ opacity: 0, y: 10 }}
              transition={{ duration: 0.2 }}
              className="mt-2 px-3"
            >
              <button
                onClick={scrollToTop}
                className={cn(
                  `text-foreground-muted flex w-full cursor-pointer items-center gap-x-1.5
                  rounded-md text-sm font-medium`,
                  'hover:text-foreground transition-colors',
                )}
                aria-label="Scroll to top of page"
              >
                <span>Scroll to top</span>
                <ArrowCircleUpIcon className="size-4.5" />
              </button>
            </motion.div>
          )}
        </AnimatePresence>
      </nav>
    );
  }

  // Mobile view - collapsible at top
  return (
    <nav
      className="border-border bg-background-elevated/50 rounded-md border backdrop-blur-sm
        xl:hidden"
      aria-label="Table of contents"
      role="navigation"
      ref={tocRef}
    >
      <button
        type="button"
        onClick={() => setIsCollapsed(!isCollapsed)}
        className={cn(
          `text-foreground flex w-full cursor-pointer items-center justify-between
          rounded-md px-4 py-3 text-sm font-semibold transition-colors`,
          'hover:bg-background-muted/50',
        )}
        aria-expanded={!isCollapsed}
        aria-controls="toc-content"
        aria-describedby="toc-description"
      >
        <span>On this page</span>
        <motion.div
          animate={{ rotate: isCollapsed ? 0 : 180 }}
          transition={{ duration: 0.2 }}
          aria-hidden="true"
        >
          <CaretDownIcon className="size-4" />
        </motion.div>
      </button>
      <span id="toc-description" className="sr-only">
        Toggle to {isCollapsed ? 'show' : 'hide'} table of contents with{' '}
        {tocItems.length} sections. Use arrow keys to navigate when expanded.
      </span>

      <AnimatePresence initial={false}>
        {!isCollapsed && (
          <motion.div
            id="toc-content"
            initial={{ height: 0, opacity: 0 }}
            animate={{ height: 'auto', opacity: 1 }}
            exit={{ height: 0, opacity: 0 }}
            transition={{ duration: 0.2, ease: 'easeInOut' }}
            className="overflow-hidden"
            role="group"
            aria-label="Page sections - use arrow keys to navigate"
          >
            <div className="p-2">
              <ul role="list" className="!my-0 list-none pl-0">
                {tocItems.map((item, index) => (
                  <li key={item.id} className="!my-0">
                    <Link
                      ref={(el) => {
                        linksRef.current[index] = el;
                      }}
                      href={`#${item.id}`}
                      onClick={(e) => handleClick(e, item.id)}
                      onKeyDown={(e) => handleKeyDown(e, item.id, index)}
                      className={cn(
                        'text-foreground-muted block rounded-md px-3 py-1.5 text-sm transition-all',
                        'hover:text-foreground',
                        item.level === 3 ? 'pl-7' : '',
                      )}
                      aria-current={
                        activeId === item.id ? 'location' : undefined
                      }
                      aria-describedby={
                        activeId === item.id
                          ? `current-section-mobile-${index}`
                          : undefined
                      }
                    >
                      <span className="line-clamp-2">{item.text}</span>
                      {activeId === item.id && (
                        <span
                          id={`current-section-mobile-${index}`}
                          className="sr-only"
                        >
                          (current section)
                        </span>
                      )}
                    </Link>
                  </li>
                ))}
              </ul>
            </div>
          </motion.div>
        )}
      </AnimatePresence>
    </nav>
  );
};
