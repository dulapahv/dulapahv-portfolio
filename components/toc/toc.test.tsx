import { render, screen, waitFor } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { useMediaQuery } from '@/hooks/use-media-query';

import { TableOfContents } from './toc';

vi.mock('@/hooks/use-media-query', () => ({
  useMediaQuery: vi.fn()
}));

vi.mock('motion/react', () => ({
  motion: {
    div: ({ children, ...props }: React.HTMLProps<HTMLDivElement>) => (
      <div {...props}>{children}</div>
    ),
    nav: ({ children, ...props }: React.HTMLProps<HTMLDivElement>) => (
      <nav {...props}>{children}</nav>
    ),
    /* eslint-disable  @typescript-eslint/no-explicit-any */
    button: ({ children, ...props }: any) => <button {...props}>{children}</button>,
    li: ({ children, ...props }: React.HTMLProps<HTMLLIElement>) => <li {...props}>{children}</li>
  },
  AnimatePresence: ({ children }: React.FragmentProps) => <>{children}</>
}));

describe('TableOfContents', () => {
  let intersectionCallback: IntersectionObserverCallback;
  let mutationCallback: MutationCallback;
  /* eslint-disable  @typescript-eslint/no-explicit-any */
  let observeIntersection: any;
  /* eslint-disable  @typescript-eslint/no-explicit-any */
  let disconnectIntersection: any;
  /* eslint-disable  @typescript-eslint/no-explicit-any */
  let observeMutation: any;
  /* eslint-disable  @typescript-eslint/no-explicit-any */
  let disconnectMutation: any;

  beforeEach(() => {
    vi.clearAllMocks();

    // Mock IntersectionObserver
    intersectionCallback = vi.fn();
    observeIntersection = vi.fn();
    disconnectIntersection = vi.fn();

    global.IntersectionObserver = class {
      constructor(callback: IntersectionObserverCallback) {
        intersectionCallback = callback;
      }
      observe = observeIntersection;
      unobserve = vi.fn();
      disconnect = disconnectIntersection;
      root = null;
      rootMargin = '';
      thresholds = [0.5];
      takeRecords = () => [];
      /* eslint-disable  @typescript-eslint/no-explicit-any */
    } as any;

    // Mock MutationObserver
    mutationCallback = vi.fn();
    observeMutation = vi.fn();
    disconnectMutation = vi.fn();

    global.MutationObserver = class {
      constructor(callback: MutationCallback) {
        mutationCallback = callback;
      }
      observe = observeMutation;
      disconnect = disconnectMutation;
      takeRecords = () => [];
      /* eslint-disable  @typescript-eslint/no-explicit-any */
    } as any;

    // Mock media query - default to desktop view
    vi.mocked(useMediaQuery).mockImplementation((query: string) => {
      if (query === '(min-width: 1350px)') return true;
      return false;
    });

    // Setup document with headings
    document.body.innerHTML = `
      <article>
        <h2 id="heading-1">Heading 1</h2>
        <h3 id="heading-2">Heading 2</h3>
        <h2 id="heading-3">Heading 3</h2>
      </article>
    `;
  });

  afterEach(() => {
    document.body.innerHTML = '';
  });

  describe('Rendering - Desktop View', () => {
    it('should render table of contents navigation', () => {
      render(<TableOfContents />);
      const nav = screen.getByRole('navigation', { name: /table of contents/i });
      expect(nav).toBeInTheDocument();
    });

    it('should render "On this page" heading', async () => {
      const { container } = render(<TableOfContents />);

      await waitFor(() => {
        const heading = container.querySelector('h2');
        expect(heading?.textContent).toBe('On this page');
      });
    });

    it('should render all headings from the document', async () => {
      render(<TableOfContents />);

      await waitFor(() => {
        expect(screen.getByRole('link', { name: /heading 1/i })).toBeInTheDocument();
        expect(screen.getByRole('link', { name: /heading 2/i })).toBeInTheDocument();
        expect(screen.getByRole('link', { name: /heading 3/i })).toBeInTheDocument();
      });
    });

    it('should have correct href for each heading', async () => {
      render(<TableOfContents />);

      await waitFor(() => {
        expect(screen.getByRole('link', { name: /heading 1/i })).toHaveAttribute(
          'href',
          '#heading-1'
        );
        expect(screen.getByRole('link', { name: /heading 2/i })).toHaveAttribute(
          'href',
          '#heading-2'
        );
        expect(screen.getByRole('link', { name: /heading 3/i })).toHaveAttribute(
          'href',
          '#heading-3'
        );
      });
    });

    it('should mark the first heading as active initially', async () => {
      render(<TableOfContents />);

      await waitFor(() => {
        const firstLink = screen.getByRole('link', { name: /heading 1/i });
        expect(firstLink).toHaveAttribute('aria-current', 'location');
      });
    });
  });

  describe('Rendering - Mobile View', () => {
    beforeEach(() => {
      vi.mocked(useMediaQuery).mockImplementation((query: string) => {
        if (query === '(min-width: 1350px)') return false;
        return false;
      });
    });

    it('should render toggle button', async () => {
      render(<TableOfContents />);

      await waitFor(() => {
        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();
        expect(button.textContent).toContain('On this page');
      });
    });

    it('should start collapsed', async () => {
      render(<TableOfContents />);

      await waitFor(() => {
        const nav = screen.getByRole('navigation');
        expect(nav).toHaveAttribute('aria-label', 'Table of contents');
      });
    });

    it('should expand when clicking toggle button', async () => {
      render(<TableOfContents />);

      await waitFor(() => {
        const button = screen.getByRole('button');
        button.click();
      });

      await waitFor(() => {
        const links = screen.getAllByRole('link');
        expect(links.length).toBeGreaterThan(0);
      });
    });
  });

  describe('Active heading tracking', () => {
    it('should have intersection observer callback defined', async () => {
      render(<TableOfContents />);

      await waitFor(() => {
        expect(screen.getByRole('link', { name: /heading 1/i })).toHaveAttribute(
          'aria-current',
          'location'
        );
      });

      // Intersection callback should be defined and ready to handle changes
      expect(intersectionCallback).toBeDefined();
    });
  });

  describe('IntersectionObserver', () => {
    it('should create IntersectionObserver', () => {
      render(<TableOfContents />);
      expect(intersectionCallback).toBeDefined();
    });

    it('should observe headings', async () => {
      render(<TableOfContents />);

      await waitFor(() => {
        expect(observeIntersection).toHaveBeenCalled();
      });
    });

    it('should disconnect observer on unmount', () => {
      const { unmount } = render(<TableOfContents />);
      unmount();
      expect(disconnectIntersection).toHaveBeenCalled();
    });
  });

  describe('MutationObserver', () => {
    it('should create MutationObserver', () => {
      render(<TableOfContents />);
      expect(mutationCallback).toBeDefined();
    });

    it('should observe article element for mutations', async () => {
      render(<TableOfContents />);

      await waitFor(() => {
        expect(observeMutation).toHaveBeenCalled();
      });
    });

    it('should disconnect mutation observer on unmount', () => {
      const { unmount } = render(<TableOfContents />);
      unmount();
      expect(disconnectMutation).toHaveBeenCalled();
    });
  });

  describe('Accessibility', () => {
    it('should have aria-label for navigation', () => {
      render(<TableOfContents />);

      const nav = screen.getByRole('navigation');
      expect(nav).toHaveAttribute('aria-label', 'Table of contents');
    });

    it('should have aria-current on active link', async () => {
      render(<TableOfContents />);

      await waitFor(() => {
        const activeLink = screen.getByRole('link', { name: /heading 1/i });
        expect(activeLink).toHaveAttribute('aria-current', 'location');
      });
    });

    it('should have screen reader instructions for desktop view', async () => {
      const { container } = render(<TableOfContents />);

      await waitFor(() => {
        const group = container.querySelector('[role="group"]');
        expect(group).toHaveAttribute('aria-label');
      });
    });
  });

  describe('Heading levels', () => {
    it('should render nested headings with proper indentation', async () => {
      document.body.innerHTML = `
        <article>
          <h2 id="h2-1">Level 2</h2>
          <h3 id="h3-1">Level 3</h3>
        </article>
      `;

      render(<TableOfContents />);

      await waitFor(() => {
        expect(screen.getByRole('link', { name: /level 2/i })).toBeInTheDocument();
        expect(screen.getByRole('link', { name: /level 3/i })).toBeInTheDocument();
      });
    });

    it('should handle headings with special characters', async () => {
      document.body.innerHTML = `
        <article>
          <h2 id="special-chars">Heading with "quotes" & symbols</h2>
        </article>
      `;

      render(<TableOfContents />);

      await waitFor(() => {
        const link = screen.getByRole('link', { name: /heading with "quotes" & symbols/i });
        expect(link).toBeInTheDocument();
      });
    });
  });

  describe('Dynamic content updates', () => {
    it('should have mutation observer set up to detect changes', async () => {
      render(<TableOfContents />);

      await waitFor(() => {
        expect(screen.getAllByRole('link')).toHaveLength(3);
      });

      // Mutation observer should be observing the article element
      expect(observeMutation).toHaveBeenCalled();
      expect(mutationCallback).toBeDefined();
    });
  });

  describe('Empty state', () => {
    it('should return null when document has no headings', () => {
      document.body.innerHTML = '<article></article>';
      render(<TableOfContents />);

      // Component returns null when there are no headings
      const nav = screen.queryByRole('navigation');
      expect(nav).toBeNull();
    });
  });

  describe('Snapshot', () => {
    it('should match snapshot - desktop view', async () => {
      const { container } = render(<TableOfContents />);

      await waitFor(() => {
        expect(screen.getAllByRole('link').length).toBeGreaterThan(0);
      });

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot - mobile view', async () => {
      vi.mocked(useMediaQuery).mockImplementation((query: string) => {
        if (query === '(min-width: 1350px)') return false;
        return false;
      });

      const { container } = render(<TableOfContents />);

      await waitFor(() => {
        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();
      });

      expect(container).toMatchSnapshot();
    });
  });
});
