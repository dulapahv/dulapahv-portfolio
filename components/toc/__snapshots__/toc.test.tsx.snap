// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableOfContents > Snapshot > should match snapshot - desktop view 1`] = `
<div>
  <nav
    aria-label="Table of contents"
    class="fixed top-18 left-[calc(100vw/2+424px)] z-40 hidden w-56 xl:block"
    role="navigation"
  >
    <h2
      class="text-foreground px-3 text-sm font-semibold"
    >
      On this page
    </h2>
    <div
      aria-label="Page sections - use arrow keys to navigate"
      class="max-h-[calc(100vh-16rem)] overflow-y-auto py-2"
      role="group"
    >
      <ul
        class="space-y-1"
        role="list"
      >
        <li
          class="mb-0"
        >
          <a
            aria-current="location"
            aria-describedby="current-section-0"
            class="group relative block rounded-md px-3 py-1 text-sm font-medium transition-all text-mirai-red"
            href="#heading-1"
          >
            <span
              aria-hidden="true"
              class="absolute top-1/2 left-0 h-full w-0.5 -translate-y-1/2 transition-all bg-mirai-red"
            />
            <span
              class="relative line-clamp-2"
            >
              Heading 1
            </span>
            <span
              class="sr-only"
              id="current-section-0"
            >
              (current section)
            </span>
          </a>
        </li>
        <li
          class="mb-0"
        >
          <a
            class="group relative block rounded-md px-3 py-1 text-sm font-medium transition-all pl-7 text-foreground-muted hover:text-foreground"
            href="#heading-2"
          >
            <span
              aria-hidden="true"
              class="absolute top-1/2 left-0 h-full w-0.5 -translate-y-1/2 transition-all bg-white/80 dark:bg-neutral-300/80"
            />
            <span
              class="relative line-clamp-2"
            >
              Heading 2
            </span>
          </a>
        </li>
        <li
          class="mb-0"
        >
          <a
            class="group relative block rounded-md px-3 py-1 text-sm font-medium transition-all text-foreground-muted hover:text-foreground"
            href="#heading-3"
          >
            <span
              aria-hidden="true"
              class="absolute top-1/2 left-0 h-full w-0.5 -translate-y-1/2 transition-all bg-white/80 dark:bg-neutral-300/80"
            />
            <span
              class="relative line-clamp-2"
            >
              Heading 3
            </span>
          </a>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;

exports[`TableOfContents > Snapshot > should match snapshot - mobile view 1`] = `
<div>
  <nav
    aria-label="Table of contents"
    class="border-border bg-background-elevated/50 rounded-md border backdrop-blur-sm"
    role="navigation"
  >
    <button
      aria-controls="toc-content"
      aria-describedby="toc-description"
      aria-expanded="false"
      class="text-foreground flex w-full cursor-pointer items-center justify-between rounded-md px-4 py-3 text-sm font-semibold transition-colors hover:bg-background-muted/50"
      type="button"
    >
      <span>
        On this page
      </span>
      <div
        animate="[object Object]"
        aria-hidden="true"
        transition="[object Object]"
      >
        <svg
          class="size-4"
          fill="currentColor"
          height="1em"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
          />
        </svg>
      </div>
    </button>
    <span
      class="sr-only"
      id="toc-description"
    >
      Toggle to 
      show
       table of contents with 
      3
       sections. Use arrow keys to navigate when expanded.
    </span>
  </nav>
</div>
`;
